File: leetcodeCpp/Longest Common Prefix/Longest Common Prefix/Longest_Common_Prefix.cpp
/*
Write a function to find the longest common prefix string amongst an array of strings.
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
	string longestCommonPrefix(vector<string>& strs) {
		string comprefix;
		if (strs.empty())
			return comprefix;
		for (int i = 0; i<strs[0].size(); i++) {
			for (int j = 1; j<strs.size(); j++) {
				if (i >= strs[j].size() || strs[j][i] != strs[0][i])
					return comprefix;
			}
			comprefix.push_back(strs[0][i]);
		}
		return comprefix;
	}
};

int main() {
	vector<string> A;
	string comprefix;

	A = { "hello","hell","hel","e" };
	
	Solution solutionobject;
	comprefix = solutionobject.longestCommonPrefix(A);

	for (unsigned i = 0; i < comprefix.size(); i++) {
		cout << comprefix[i] << endl;
	}
	system("pause");
	return 0;
}

File: leetcodeGolang/longestCommonPrefix.go
/*
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: ["flower","flow","flight"]
Output: "fl"
Example 2:

Input: ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
Note:

All given inputs are in lowercase letters a-z.
*/

package main

import (
	"fmt"
)

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	res := strs[0]
	for _, str := range strs {
		if len(str) < len(res) {
			res = str
		}
	}

	for i, alph := range res {
		for j := 0; j < len(strs); j++ {
			if strs[j][i] != byte(alph) {
				return strs[j][:i]
			} 
		}
	}
	return res
}

func main()  {
	a := []string{"flowers", "flow", "flight"}
	b := longestCommonPrefix(a)
	fmt.Println(b)

}